### PR TITLE
feat: delete proteus sessions FS-1595

### PR DESCRIPTION
**Pull request by johnxnguyen**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

In this PR we port over the final methods from Cryptobox to CoreCrypto that enable the deleting of proteus sessions.

Also in this PR we try to eliminate the dependency on the keystore in favor of using the `ProteusProvider`. Note that there are still some places where we use the keystore and not the proteus service, such as when migrating session ids. We'll address this in a future PR.

One of consequences of removing the keystore from the user client is that we no longer fetch the self client to access the keystore, instead we take it directly from the managed object context. I don't know why we were always accessing the keystore from the self user client, since it is just an indirection to the same managed object context available in scope. 

#### Test Coverage 

- Several tests have been updated for API changes (e.g removing the keystore from the user client)
- New tests for session deletion and various fingerprints assertions are created for the `ProteusService` path, existing tests have been marked as legacy
- New unit tests for `ProteusService` session deletion.

#### Notes
Apologies for the mixture of changes here, this PR started out as one thing but evolved into another.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift [hunk 1]
```diff
@@ -228,25 +228,30 @@ extension GenericMessage {
...
-        } else {
-            encryptedData = legacyEncrypt(
-                for: recipients,
-                with: missingClientsStrategy,
-                externalData: externalData,
-                useQualifiedIdentifiers: useQualifiedIdentifiers,
-                in: context
-            )
-        }
+        ProteusProvider(context: context).perform(
```
Comment (David-Henner) [inline wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift]:
```suggestion
        context.proteusProvider.perform(
```

File: wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift [hunk 2]
```diff
@@ -267,23 +272,28 @@ extension GenericMessage {
...
-            )
-        } else {
-            encryptedData = legacyEncrypt(
-                for: messageRecipients,
-                with: missingClientsStrategy,
-                useQualifiedIdentifiers: useQualifiedIdentifiers,
-                in: context
-            )
-        }
+        ProteusProvider(context: context).perform(
```
Comment (David-Henner) [inline wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift]: same suggestion

File: wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift [hunk 3]
```diff
@@ -301,23 +311,28 @@ extension GenericMessage {
...
-            )
-        } else {
-            encryptedData = legacyEncrypt(
-                for: recipients,
-                with: missingClientsStrategy,
-                useQualifiedIdentifiers: useQualifiedIdentifiers,
-                in: context
-            )
-        }
+        ProteusProvider(context: context).perform(
```
Comment (David-Henner) [inline wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift]: 👀

File: wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift [hunk 4]
```diff
@@ -750,26 +766,32 @@ extension GenericMessage {
...
-            return externalGenericMessage.legacyEncrypt(
-                for: recipients,
-                with: missingClientsStrategy,
-                externalData: data,
-                useQualifiedIdentifiers: useQualifiedIdentifiers,
-                in: context
-            )
-        }
-
+        ProteusProvider(context: context).perform(
```
Comment (David-Henner) [inline wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift]: 👀

File: wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift (hunk 2 above)
Comment (netbe) [inline wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift]:
```suggestion
        context.proteusProvider.perform(
```

File: wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift (hunk 3 above)
Comment (netbe) [inline wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift]:
```suggestion
        context.proteusProvider.perform(
```

File: wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift (hunk 4 above)
Comment (netbe) [inline wire-ios-data-model/Source/Model/Message/ZMClientMessage+Encryption.swift]:
```suggestion
        context.proteusProvider.perform(
```
just to make it consistent across the code how we get the provider

File: wire-ios-data-model/Source/Model/UserClient/UserClient.swift [hunk 5]
```diff
@@ -649,23 +660,56 @@ public extension UserClient {
...
+    func deleteSession(context: NSManagedObjectContext? = nil) throws {
+        guard
+            !isSelfClient(),
+            let context = context ?? managedObjectContext,
+            let sessionID = proteusSessionID
+        else {
+            return
+        }
+
+        try ProteusProvider(context: context).perform(
```
Comment (netbe) [inline wire-ios-data-model/Source/Model/UserClient/UserClient.swift]:
```suggestion
        try context.proteusProvider.perform(
```

Comment (David-Henner) [inline wire-ios-data-model/Source/Model/UserClient/UserClient.swift]: 👀

File: wire-ios-data-model/Tests/Source/Model/UserClient/UserClientTests.swift [hunk 6]
```diff
@@ -284,15 +354,59 @@ final class UserClientTests: ZMBaseManagedObjectTest {
     }
 
     func testThatItRefetchesMissingFingerprintForUserWithSession() {
-        // given
+        // Given
+        var flag = DeveloperFlag.proteusViaCoreCrypto
+        flag.isOn = true
```
Comment (netbe) [inline wire-ios-data-model/Tests/Source/Model/UserClient/UserClientTests.swift:359]: maybe we should add an helper method to make sure we always flip back the switch like:
```
extension DeveloperFlag {
    mutating func toggle(_ enabled: Bool, block: () -> Void) {
        self.isOn = enabled
        block()
        self.isOn = !enabled
    }
}
```

**Review by David-Henner — APPROVED** [review]

File: wire-ios-data-model/Tests/Source/Model/UserClient/UserClientTests.swift (hunk 6 above)
Comment (johnxnguyen) [reply wire-ios-data-model/Tests/Source/Model/UserClient/UserClientTests.swift:359]: I was thinking about a change to the `DeveloperFlag` storage so we don't actually use `UserDefaults` in test but just have some in memory storage that we reset in test setup methods. In any case, I think we can do these kind of improvements separately.

**Review by netbe — APPROVED** [review]

**Pull request merged**